handlers/viber/viber.go

```diff
@@ -10,3 +27,280 @@ POST /handlers/viber_public/uuid?sig=sig
+		}
+
+		// build our msg
+		msg := h.Backend().NewIncomingMsg(channel, urn, text).WithExternalID(fmt.Sprintf("%d", viberMsg.MessageToken)).WithContactName(contactName)
+
+		if mediaURL != "" {
+			msg.WithAttachment(mediaURL)
+		}
+
+		// and finally queue our message
+		err = h.Backend().WriteMsg(msg)
+		if err != nil {
+			return nil, err
+		}
+
+		return []courier.ReceiveEvent{msg}, courier.WriteMsgSuccess(w, r, []courier.Msg{msg})
+	}
+
+	return nil, courier.WriteError(w, r, fmt.Errorf("not handled, unknown event: %s", event))
+}
+
+type viberMessage struct {
+	Event        string `json:"event" validate:"required"`
+	Timestamp    int64  `json:"timestamp" validate:"required"`
+	MessageToken int64  `json:"message_token" validate:"required"`
+	Sender       struct {
+		ID   string `json:"id"`
+		Name string `json:"name"`
+	} `json:"sender"`
+	User struct {
+		ID   string `json:"id"`
+		Name string `json:"name"`
+	} `json:"user"`
+	Message struct {
+		Text    string `json:"text"`
+		Media   string `json:"media"`
+		Contact struct {
+			Name        string `json:"name"`
+			PhoneNumber string `json:"phone_number"`
+		}
+		Location struct {
+			Latitude  float64 `json:"lat"`
+			Longitude float64 `json:"lon"`
+		}
+		Type         string `json:"type"`
+		TrackingData string `json:"tracking_data"`
+	} `json:"message"`
+}
+
+// see https://developers.viber.com/docs/api/rest-bot-api/#callbacks
+func (h *handler) validateSignature(channel courier.Channel, r *http.Request) error {
+	actual := r.Header.Get(viberSignatureHeader)
+	if actual == "" {
+		return fmt.Errorf("missing request signature")
+	}
+
+	confAuth := channel.ConfigForKey(courier.ConfigAuthToken, "")
+	authToken, isStr := confAuth.(string)
+	if !isStr || authToken == "" {
+		return fmt.Errorf("invalid or missing auth token in config")
+	}
+
+	// read our body
+	body, err := ioutil.ReadAll(r.Body)
+	r.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+	expected, err := viberCalculateSignature(authToken, body)
+	if err != nil {
+		return err
+	}
+
+	// compare signatures in way that isn't sensitive to a timing attack
+	if !hmac.Equal(expected, []byte(actual)) {
+		return fmt.Errorf("invalid request signature")
+	}
+
+	return nil
+}
+
+func viberCalculateSignature(authToken string, contents []byte) ([]byte, error) {
+
+	var buffer bytes.Buffer
+	buffer.Write(contents)
+
+	// hash with SHA256
+	mac := hmac.New(sha256.New, []byte(authToken))
+	mac.Write(buffer.Bytes())
+	hash := mac.Sum(nil)
+
+	return hash, nil
+}
+
+// SendMsg sends the passed in message, returning any error
+func (h *handler) SendMsg(msg courier.Msg) (courier.MsgStatus, error) {
+	confAuth := msg.Channel().ConfigForKey(courier.ConfigAuthToken, "")
+	authToken, isStr := confAuth.(string)
+	if !isStr || authToken == "" {
+		return nil, fmt.Errorf("invalid auth token config")
+	}
+
+	viberMsg := viberOutgoingMessage{
+		AuthToken:    authToken,
+		Receiver:     msg.URN().Path(),
+		Text:         courier.GetTextAndAttachments(msg),
```

Comment:
Whats the max message length? Does viber not support attachments?

Reply:
I added the max length, for attachement we can easily support picture as attachment but for video we need to specify the size of the attachment https://developers.viber.com/docs/api/rest-bot-api/#video-message 
I think we need first a way to get that do we fetch first the mediaURL to get the size?